taichi/transforms/handle_external_ptr_boundary.cpp

```diff
@@ -0,0 +1,132 @@
+#include "taichi/ir/ir.h"
+#include "taichi/ir/statements.h"
+#include "taichi/ir/transforms.h"
+#include "taichi/ir/visitors.h"
+#include "taichi/transforms/check_out_of_bound.h"
+#include "taichi/transforms/utils.h"
+#include <set>
+
+namespace taichi::lang {
+
+class HandleExternalPtrBound : public BasicStmtVisitor {
+ public:
+  using BasicStmtVisitor::visit;
+  std::set<int> visited;
+  DelayedIRModifier modifier;
+
+  bool is_done(Stmt *stmt) {
+    return visited.find(stmt->instance_id) != visited.end();
+  }
+
+  void set_done(Stmt *stmt) {
+    visited.insert(stmt->instance_id);
+  }
+
+  void visit(ExternalPtrStmt *stmt) override {
+    if (stmt->boundary == BoundaryMode::kUnsafe || is_done(stmt))
+      return;
+    if (stmt->boundary == BoundaryMode::kClamp) {
+      auto new_stmts = VecStatement();
+      auto zero = new_stmts.push_back<ConstStmt>(TypedConstant(0));
+
+      int flattened_element = 1;
+      for (int i = 0; i < stmt->element_shape.size(); i++) {
+        flattened_element *= stmt->element_shape[i];
+      }
+      for (int i = 0; i < stmt->indices.size(); i++) {
+        auto lower_bound = zero;
+        auto check_lower_bound = new_stmts.push_back<BinaryOpStmt>(
```

Comment:
`lower_bound_checked` or simply `checked_index` would be a better name for the variable.